### DISCUSSION TITLE
Encapsulamento de variável de classe abstrata.

### DISCUSSION THREAD

**Discussion by brunos0** [Flutterando/forum#65] [Perguntas e Respostas]
Bom dia Galera.

Estou aprendendo Dart e estou estudando no momento sobre Polimorfismo e bateu uma dúvida. Criei uma classe "Modelo" Abstrata, declarei uma variável com um underscore " _ " e isso em classes convencionais tornam a variável "private", só sendo possíveis acessar o conteúdo delas dentro da classe que a originou. Corrijam me se estiver enganado, mas isso seria o tal do encapsulamento, que daria uma maior segurança na leitura e na manipulação das variáveis.
Criei também uma classe "Especializada" com o uso da palavra "implements" apontando a classe modelo. Dessa forma eu tive que definir os métodos gets, sets, e os demais métodos. Quando criei a função main e instanciei um objeto do tipo "Modelo" com o construtor da classe "Especializada", eu consegui acessar o conteúdo diretamente da variável "private" e manipular o conteúdo da variável sem passar pelos gets/sets. Venho de uma linguagem totalmente procedural então estou apanhando um pouco com o Dart, mas estou achando muito maneiro a linguagem. :D 

Outra coisa que achei estranho foi: na classe "Especializada", criei uma variável "private" e um método exclusivo. Quando instanciei o objeto do jeito acima, não consegui acessar nem o método nem a variável criada. Só consegui acessar a variável e o método, chamado o nome da classe como se fosse uma função e colocando o sinal de ponto " . " juntamente com o nome da variável (que eu também acho que não deveria ser possível acessar pois ela também tem o _ no nome) ou o nome do método.

Ultima coisinha ( eu prometo ). Tentei especificar o método2 na classe "Especializada", mas tentando aproveitar uma eventual operação da classe "Modelo", mas não consegui, nem usando a palavra @override  com a super.executarMetodo2() ou Xpto.executarMetodo2(). 

Esse comportamento está correto? Se sim, como encapsular as variáveis de classes abstratas/classes que implementam classes abstratas? Como posso acessar variáveis/métodos das classes "Especialializadas"?

Vou deixar o código que eu fiz para vocês darem pitaco de como eu posso ajustar meu código:

void main (){
  
  Xpto xyz = SeiLa();
  xyz.executarMetodo1();
  xyz.executarMetodo2();
  print(xyz._valor1);
  xyz._valor1 = 34 ; 
  print(xyz._valor1);
 
}

abstract class Xpto{

  late int    _valor1;
  late String _texto1;

  int get gval1;
  String get gtxt1;
  void set sval1 (int a);
  void set stxt1 (String b);

  void executarMetodo1();
  void executarMetodo2(){ print("Xpto metodo2"); }
}

class SeiLa implements Xpto{
  
  int     _valor1   = 10;
  String  _texto1   = "X";
  bool    _logic1 = false;
  
  int get gval1 {
    return this._valor1; 
  }
  String get gtxt1 {
    return this._texto1; 
  }
  
  set sval1 (int a){
    this._valor1 = a; 
  }

  set stxt1 (String b){
    this._texto1 = b;
  }
  
  void executarMetodo1(){
    print("Seila metodo1");
  }

  void executarMetodo2(){
    print("Seila metodo2");
  }
}

Obrigado por tirar essas pequenas dúvidas desse pobre Padawan.
Bruno$0

**Comment by mateusfccp** [comment]
Olá, @brunos0.

Primeiramente, quando for postar código, por favor, use a formatação para que fique fácil de entender:

```dart
void main() {
  Xpto xyz = SeiLa();
  xyz.executarMetodo1();
  xyz.executarMetodo2();
  print(xyz._valor1);
  xyz._valor1 = 34;
  print(xyz._valor1);
}

abstract class Xpto {
  late int _valor1;
  late String _texto1;

  int get gval1;
  String get gtxt1;
  void set sval1(int a);
  void set stxt1(String b);

  void executarMetodo1();
  void executarMetodo2() {
    print("Xpto metodo2");
  }
}

class SeiLa implements Xpto {
  int _valor1 = 10;
  String _texto1 = "X";
  bool _logic1 = false;

  int get gval1 {
    return this._valor1;
  }

  String get gtxt1 {
    return this._texto1;
  }

  set sval1(int a) {
    this._valor1 = a;
  }

  set stxt1(String b) {
    this._texto1 = b;
  }

  void executarMetodo1() {
    print("Seila metodo1");
  }

  void executarMetodo2() {
    print("Seila metodo2");
  }
}
```

Vamos às dúvidas.

---

> Corrijam me se estiver enganado, mas isso seria o tal do encapsulamento, que daria uma maior segurança na leitura e na manipulação das variáveis.

Pode ser, ou não. Encapsulamento pode ser entendido de duas forma:

* Um mecanismo de linguagem para restringir o acesso direto a alguns componentes; ou
* Um construto da linguagem que facilite o agrupamento de datas com métodos.

Nesse caso, você está se referindo ao primeiro entendimento, o que está certo.

---

> Quando criei a função main e instanciei um objeto do tipo "Modelo" com o construtor da classe "Especializada", eu consegui acessar o conteúdo diretamente da variável "private" e manipular o conteúdo da variável sem passar pelos gets/sets.

O conceito de "privado", em Dart, diz respeito à privacidade dentro de uma biblioteca. O escopo da biblioteca é o de um só arquivo. Isso significa que, se a definição do nome (`_nome`) estiver no mesmo arquivo, você pode acessá-lo independente de estar ou não na classe que o definiu.

Observe a seção 6.2 da [especificação atual da linguagem Dart](https://dart.dev/guides/language/specifications/DartLangSpec-v2.10.pdf):

> Dart supports two levels of privacy: public and private. A declaration is private iff its name is private, otherwise it is public. A name q is private iff any one of the identifiers that comprise q is private, otherwise it is public. An identifier is private iff it begins with an underscore (the _ character) otherwise it is public.

> A declaration m is accessible to a library L if m is declared in L or if m is public.

> This means private declarations may only be accessed within the library in which they are declared.

> Privacy applies only to declarations within a library, not to the library declaration as a whole. This is because libraries do not reference each other by name, and so the idea of a private library is meaningless (19.1). Thus, if the name of a library begins with an underscore, it has no effect on the accessibility of the library or its members.

> Privacy is, at this point, a static notion tied to a particular piece of code (a library). It is designed to support software engineering concerns rather than security concerns. Untrusted code should always run in an another isolate.

> Privacy is indicated by the name of a declaration—hence privacy and naming are not orthogonal. This has the advantage that both humans and machines can recognize access to private declarations at the point of use without knowledge of the context from which the declaration is derived.

---

> Outra coisa que achei estranho foi: na classe "Especializada", criei uma variável "private" e um método exclusivo. Quando instanciei o objeto do jeito acima, não consegui acessar nem o método nem a variável criada. Só consegui acessar a variável e o método, chamado o nome da classe como se fosse uma função e colocando o sinal de ponto " . " juntamente com o nome da variável (que eu também acho que não deveria ser possível acessar pois ela também tem o _ no nome) ou o nome do método.

Não entendi bem o que você quis dizer.

---

> Ultima coisinha ( eu prometo ). Tentei especificar o método2 na classe "Especializada", mas tentando aproveitar uma eventual operação da classe "Modelo", mas não consegui, nem usando a palavra @OverRide com a super.executarMetodo2() ou Xpto.executarMetodo2().

A classe `SeiLa` implementa a classe `Xpto`. Isso significa que você está apenas utilizando a **interface**, ou seja, o contrato, estabelecido pela classe `Xpto`. Por isso, você não consegue usar `super.method()`. Para estender o **comportamento** da classe `Xpto`, em vez de usar `implements`, você deve usar `extends`.

Segue um link com mais detalhes sobre esse assunto: https://medium.com/@manoelsrs/dart-extends-vs-implements-vs-with-96abd515f04e
